packages/@react-spectrum/cards/src/GridLayout.tsx

```diff
@@ -0,0 +1,283 @@
+    } else {
+      // The approach from v2 uses indexes where other v3 layouts iterate through every node/root node. This feels more efficient
+      let firstVisibleItem = this.getIndexAtPoint(rect.x, rect.y);
+      let lastVisibleItem = this.getIndexAtPoint(rect.maxX, rect.maxY);
+
+      for (let index = firstVisibleItem; index < lastVisibleItem; index++) {
+        let keyFromIndex = this.collection.at(index).key;
+        let layoutInfo = this.layoutInfos.get(keyFromIndex);
+        if (this.isVisible(layoutInfo, rect)) {
+          res.push(layoutInfo);
+        }
+      }
+
+      // Check if loader is in view and add to res if so
+      let loader = this.layoutInfos.get('loader');
+      if (loader && this.isVisible(loader, rect)) {
+        res.push(loader);
+      }
+    }
+
+    return res;
+  }
+
+  validate() {
+    this.collection = this.virtualizer.collection;
+    this.buildCollection();
+
+    // Remove layout info that doesn't exist in new collection
+    if (this.lastCollection) {
+      for (let key of this.lastCollection.getKeys()) {
+        if (!this.collection.getItem(key)) {
+          this.layoutInfos.delete(key);
+        }
+      }
+
+      // TODO: dunno why I had to add this but it gets rid of the loading spinner, otherwise it remains on initial async load
+      // Can't figure out why ListLayout didn't need that
+      if (!this.isLoading) {
+        this.layoutInfos.delete('loader');
+      }
+    }
+
+    this.lastCollection = this.collection;
+  }
+
+  buildCollection() {
+    let visibleWidth = this.virtualizer.visibleRect.width;
+    let visibleHeight = this.virtualizer.visibleRect.height;
+
+    // Compute the number of rows and columns needed to display the content
+    let availableWidth = visibleWidth - this.margin * 2;
+    let columns = Math.floor(availableWidth / (this.minItemSize.width + this.minSpace.width));
+    this.numColumns = Math.max(1, Math.min(this.maxColumns, columns));
+    this.numRows = Math.ceil(this.collection.size / this.numColumns);
+
+    // Compute the available width (minus the space between items)
+    let width = availableWidth - (this.minSpace.width * Math.max(0, this.numColumns - 1));
+
+    // Compute the item width based on the space available
+    let itemWidth = Math.floor(width / this.numColumns);
+    itemWidth = Math.max(this.minItemSize.width, Math.min(this.maxItemSize.width, itemWidth));
+
+    // Compute the item height, which is proportional to the item width
+    let t = ((itemWidth - this.minItemSize.width) / this.minItemSize.width);
+    let itemHeight = this.minItemSize.height + this.minItemSize.height * t;
+    itemHeight = Math.max(this.minItemSize.height, Math.min(this.maxItemSize.height, itemHeight)) + this.itemPadding;
+
+    this.itemSize = new Size(itemWidth, itemHeight);
+
+    // Compute the horizontal spacing and content height
+    this.horizontalSpacing = Math.floor((visibleWidth - this.numColumns * this.itemSize.width) / (this.numColumns + 1));
```

Comment:
This feels like it should be `this.horizontalSpacing = this.numColumns < 2 ? 0 : Math.floor((availableWidth - this.numColumns * this.itemSize.width) / (this.numColumns - 1));`
This way from v2 will unfortunately make uneven margins around the CardView for some specific case (e.g. large custom margin with constraints on maxColumn)

Runs into some issues with being off 1 px in the test and with getVisibleLayoutInfos not loading the last item that should be in view, TODO: investigate.